### DISCUSSION TITLE
Help understanding gap in uniform buffer size/offsets?

### DISCUSSION THREAD

**Discussion by OmarShehata** [gpuweb/gpuweb#2348] [Q&A] [answered]
I have a compute shader, and I want to attach the following data in a uniform buffer object:

```glsl
[[block]] struct UBO {
  screenWidth: f32;
  screenHeight: f32;
  matrix: mat4x4<f32>;
};
```

### Question 1 - what should the size of my UBO be? 

This is how I'm computing it:

```javascript
const UBOBufferSize =
    4 * 16 + // 4x4 matrix
    4 * 2// screen width & height

const UBOBuffer = device.createBuffer({
   size: UBOBufferSize,
   usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
});

/// ... in the bind group layout ...
{
  binding: 1,// uniform buffer
  visibility: GPUShaderStage.COMPUTE,
  buffer: {
    type: "uniform",
  },
}
```

But this gives me an error that the size is incorrect:

```
Binding sizes are too small for bind group [BindGroup] at index 0
```

I can fix it by adding a `+ 8` to my `UBOBufferSize`, but I don't understand why that's necessary?

### Question 2 - how much do I need to offset when writing the width/height followed by the matrix?

I assume the answer to this is the same as above. When trying to write data to the buffer, I would think I can wrote the width/height at offset 0, then the 16 matrix value at offset 8 (since a float32 is 4 bytes).

But this does NOT work. In the compute shader the value of `matrix[0][0]` will be 3, not the first element 1 as expected. I need to write the matrix values with an offset of 16 instead for all the uniforms in my compute shader to have the correct value. 

```javascript
import { mat4 } from 'gl-matrix';
/// ...
const matrix = mat4.create();
mat4.set(matrix, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16);

const uniformData = [WIDTH, HEIGHT];
const uniformTypedArray = new Float32Array(uniformData);
device.queue.writeBuffer(
  UBOBuffer, 0, uniformTypedArray.buffer,
  uniformTypedArray.byteOffset, uniformTypedArray.byteLength
 );
device.queue.writeBuffer(
  UBOBuffer, 8, matrix.buffer, /////// this does NOT work. Has to be 16 instead?
  matrix.byteOffset, matrix.byteLength
  )
```

So it seems like there's some 8 byte gap between the width/height and the matrix? Is this expected?

**Comment by ben-clayton** [comment, marked as answer]
The layout rules are described in detail here:
https://www.w3.org/TR/WGSL/#alignment-and-size

Scalars (`f32`) are 4-byte sized and aligned.
`mat4x4` is 64-byte sized and 16-byte aligned.

So this is the layout:

```rust
[[block]] struct UBO {
  screenWidth: f32;      // offset: 0, size: 4
  screenHeight: f32;     // offset: 4, size: 4
  // <implicit-padding>  // offset: 8, size: 8
  matrix: mat4x4<f32>;   // offset: 16, size: 64
};
```

The size of UBO is 80 bytes:

> The size of a structure is equal to the offset plus the size of its last member, rounded to the next multiple of the structure’s alignment:

> SizeOf(S) = roundUp(AlignOf(S), OffsetOfMember(S, L) + SizeOfMember(S, L))
Where L is the last member of the structure

You could move the matrix to be the first member of the struct to remove the padding between the fields, but UBO would still be 80 bytes, as the alignment of the structure is equal to the member with the largest alignment (16 bytes in this case).

**Reply by OmarShehata** [reply]
Thank you so much @ben-clayton !! I hadn't been thinking of alignment at all so far. I appreciate the links and the very well written explanation!
